### PR TITLE
Staking Reparameterization and Datatrust Registration

### PR DESCRIPTION
**Pull request by rbharath**
An experimental implementation of #152. More details once ready to review

### CODE REVIEW

File: tests/creatorblocktest/scenariofive/deploy_scenario_five_test.go [hunk 1]
```diff
@@ -101,6 +101,16 @@ func TestMain(m *testing.M) {
 		CostPerByte: big.NewInt(test.ONE_GWEI * 100),
 	})
 
+	// auth backend will need at least the stake
```
Comment (rbharath) [inline tests/creatorblocktest/scenariofive/deploy_scenario_five_test.go]: Transfer funds to the backend address to enable staking.

File: tests/creatorblocktest/scenariofour/deploy_scenario_four_test.go [hunk 2]
```diff
@@ -67,17 +67,27 @@ func TestMain(m *testing.M) {
 		CostPerByte: big.NewInt(test.ONE_GWEI * 100),
 	})
 
+	// auth backend will need at least the stake
```
Comment (rbharath) [inline tests/creatorblocktest/scenariofour/deploy_scenario_four_test.go]: Again transferring funds to backend address.

File: tests/votingtest/voting_functionality_test.go [hunk 3]
```diff
@@ -11,46 +11,57 @@ import (
...
+	// Check that there is not a candidate now
+	isCan, _ := deployed.VotingContract.IsCandidate(nil, reparamHash)
+	if isCan {
+		t.Errorf("Expected isCandidate to be false before reparam submission, got: %v", isCan)
+	}
+
 	// we need a candidate, a privileged contract must add it...
 	_, candErr := deployed.ParameterizerContract.Reparameterize(test.GetTxOpts(context.AuthUser1, nil,
-		big.NewInt(test.ONE_GWEI*2), 200000), test.VOTE_BY, big.NewInt(99))
+		big.NewInt(test.ONE_GWEI*2), 1000000), test.VOTE_BY, big.NewInt(100000))
```
Comment (rbharath) [inline tests/votingtest/voting_functionality_test.go]: `Parameterizer.Reparameterize` needs more gas now so we have to bump the gas

**Comment by rbharath** [comment]
It turns out the changes to the contracts needed for staking in this fashion are pretty clean. I think this set of changes is ready for review.

The main changes are to the tests, where datatrusts and would-be reparameterizers need to be given funds in order to stake.

CC @robrobbins

File: contracts/Datatrust.vy [hunk 4]
```diff
@@ -24,6 +24,7 @@ contract Voting:
   def removeCandidate(hash: bytes32): modifying
   def didPass(hash: bytes32, plurality: uint256) -> bool: constant
   def pollClosed(hash: bytes32) -> bool: constant
+  def transferStake(hash: bytes32, addr: address): modifying
```
Comment (robrobbins) [inline contracts/Datatrust.vy]: i think you misunderstand the use case for `transferStake`. It's used in challenges to transfer the stake from a loser to a winner by the listing contract.

File: contracts/Voting.vy [hunk 5]
```diff
@@ -207,7 +207,8 @@ def transferStake(hash: bytes32, addr: address):
   @param hash The Candidate identifier
   @param addr The Address recieving the credit
   """
-  assert msg.sender == self.listing_address # only the listing contract will call this
+  # The caller must be privileged 
+  assert self.hasPrivilege(msg.sender)
```
Comment (robrobbins) [inline contracts/Voting.vy]: no. leave it listing as per comments above

File: tests/creatorblocktest/scenariofour/deploy_scenario_four_test.go (hunk 2 above)
Comment (robrobbins) [reply tests/creatorblocktest/scenariofour/deploy_scenario_four_test.go]: i'm ok with most the tests using transfer to get CMT to a datatrust owner

but

at least once we should mimic the real world path a non-computable datatrust will have to take in order to have something to stake with

have the datatrust call support etc... just to make sure we didn't create some weird chicken vs egg scenario

File: contracts/Voting.vy [hunk 6]
```diff
@@ -126,7 +126,7 @@ def addCandidate(hash: bytes32, kind: uint256, owner: address, stake: wei_value,
   """
   assert self.hasPrivilege(msg.sender)
   assert self.candidates[hash].owner == ZERO_ADDRESS
-  if kind == CHALLENGE: # a challenger must successfully stake a challenge
+  if kind != APPLICATION: # Only listing candidates don't require staking
```
Comment (robrobbins) [inline contracts/Voting.vy]: this is the only line of code that needed to be changed in the contracts.

just this.

i _think_ (could be missing something) the rest of the system will just work if we leave it alone.

* owner must stake
* at some point owner must call unstake if they want it back

Comment (rbharath) [reply contracts/Voting.vy]: I thought about this, and I think you're right. I'll clean up and remove the `transferStake` operations

File: contracts/Datatrust.vy (hunk 4 above)
Comment (rbharath) [reply contracts/Datatrust.vy]: Yeah, good point. I misunderstood `transferStake`. Removed

File: contracts/Voting.vy (hunk 5 above)
Comment (rbharath) [reply contracts/Voting.vy]: Switched back

**Comment by rbharath** [comment]
@robrobbins Cleaned up the unnecessary use of `transferStake` and added a test that uses the real-world flow of gaining `MarketToken` via `support`. This is ready for review again

**Pull request merged**